JS tooling: upgrade to babel 7

davkal opened the pull request:
* upgrade from babel 6 to 7
* use babel/preset-env and `useBuiltIns` to govern what is included from babel polyfill
* replace ts-loader from hot-reload webpack config, TS compilation now being done by babel plugin

Seems to work ok for `yarn build` and `yarn dev`. But with `yarn start` I get errors when accessing a dashboard: `TypeError: Cannot read property 'meta' of undefined` which seems to come from `dashnav.ts`:
```js
  /** @ngInject */
  constructor(private $scope, private dashboardSrv, private $location, public playlistSrv) {
    appEvents.on('save-dashboard', this.saveDashboard.bind(this), $scope);

    if (this.dashboard.meta.isSnapshot) {
```
Somehow the instantiation is not working (`dashboard` is missing, which is supplied via an angular directive).

davkal committed:
WIP babel 7

torkelo commented:
if dashboard is null there is could be that somehow pre assign bindings is not enabled .

https://github.com/grafana/grafana/blob/master/public/app/app.ts#L70 

not sure why hot reload should cause issues with this binding flag

davkal commented:
The bundle implications are negligible:

Before:
```
     vendor.d9d56360a977b6c2316b.js    3.39 MiB       0  [emitted]  [big]  vendor
```

After:
```
     vendor.6af89965262332b9c799.js    3.38 MiB       0  [emitted]  [big]  vendor
```

dprokop commented on scripts/webpack/webpack.hot.js:
```diff
@@ -47,25 +47,31 @@ module.exports = merge(common, {
     rules: [
       {
-        test: /\.tsx?$/,
+        test: /\.(j|t)sx?$/,
```
@davkal what's the reasoning behind this change?

davkal replied:
I wanted the config to be the same as in react-hot-reloader README in case I overlooked something. `jsx?` won't be included in this PR. Thanks for highlighting it out though.

torkelo commented:
This was tricky.

Somehow babel is doing stuff that is not compatible or the same as typescript output. Babel is setting clas properties to undefined. which breaks angularjs bindings that pre assigns properties before calling constructor. 

Possibly related issues
https://github.com/babel/babel/issues/7644
https://github.com/babel/babel/issues/7074
https://github.com/babel/babel/issues/5854
https://github.com/babel/babel/pull/8682

Not sure how to fix this. Tried some suggestions in those issues without success.

torkelo committed:
fixed issue with babel plugin proposal class properties that initiated
properties to void 0. This breaks angularjs preAssignBinding which
applies bindings to this before constructor is called. Fixed by
using fork of babel plugin.

https://github.com/babel/babel/issues/8417

torkelo committed:
Merge branch 'master' into davkal/babel-7

torkelo commented:
Pushed a fix using fork of the problematic babel plugin,
https://github.com/grafana/grafana/pull/13873/commits/fd71abc301256a133341a18348bb913c6498e2ec 

Also investigated the why the type warnings show up (Bug in webpack where there currently is no workaround other than filtering them out), they are filtered out in normal webpack watch mode using stats warningFilter, but this seems to be ignored by webpack dev server:
https://github.com/webpack/webpack-dev-server/issues/1557

davkal committed:
Stick to .tsx? for babel file test

davkal commented:
LGTM now.
